modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,799 @@
+                Paths.get(databaseRelativePath(pdsFolderName), cacheDir.getName()).toString());
+
+            try {
+                if (log.isInfoEnabled())
+                    log.info("Copying files of the cache group [from=" + snpCacheDir + ", to=" + cacheDir + ']');
+
+                for (File snpFile : snpCacheDir.listFiles()) {
+                    if (stopChecker.getAsBoolean())
+                        return;
+
+                    File target = new File(cacheDir, snpFile.getName());
+
+                    if (log.isDebugEnabled()) {
+                        log.debug("Copying file from the snapshot " +
+                            "[snapshot=" + opCtx.snpName +
+                            ", src=" + snpFile +
+                            ", target=" + target + "]");
+                    }
+
+                    Files.copy(snpFile.toPath(), target.toPath());
+                }
+            }
+            catch (IOException e) {
+                throw new IgniteCheckedException("Unable to copy file [snapshot=" + opCtx.snpName +
+                    ", grp=" + FilePageStoreManager.cacheGroupName(cacheDir) + ']', e);
+            }
+        }
+    }
+
+    /**
+     * Rollback changes made by process in specified cache group.
+     *
+     * @param opCtx Snapshot restore operation context.
+     */
+    private void rollback(@Nullable SnapshotRestoreContext opCtx) {
+        if (opCtx == null || F.isEmpty(opCtx.dirs))
+            return;
+
+        if (log.isInfoEnabled())
+            log.info("Performing local rollback routine for restored cache groups [requestID=" + opCtx.reqId + ']');
+
+        try {
+            for (File cacheDir : opCtx.dirs) {
+                if (!cacheDir.exists())
+                    continue;
+
+                if (log.isInfoEnabled())
+                    log.info("Cleaning up directory " + cacheDir);
+
+                U.delete(cacheDir);
+            }
+        }
+        catch (Exception e) {
+            log.error("Failed to perform rollback [requestID=" + opCtx.reqId + ", snapshot=" + opCtx.snpName + ']', e);
+        }
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @param res Results.
+     * @param errs Errors.
+     */
+    private void finishPrepare(UUID reqId, Map<UUID, ArrayList<StoredCacheData>> res, Map<UUID, Exception> errs) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (ctx.clientNode())
+            return;
+
+        Exception failure = checkFailure(errs, opCtx0, res.keySet());
+
+        if (failure == null) {
+            assert opCtx0 != null : ctx.localNodeId();
+
+            Map<Integer, StoredCacheData> globalCfgs = new HashMap<>();
+
+            for (List<StoredCacheData> storedCfgs : res.values()) {
+                if (storedCfgs == null)
+                    continue;
+
+                for (StoredCacheData cacheData : storedCfgs)
+                    globalCfgs.put(CU.cacheId(cacheData.config().getName()), cacheData);
```

Comment:
I think you can put directly to `opCtx0.cfgs`, use the `putIfAbsent` instead and enrich the configurations that you already have on the local node, isn't it?

Reply:
We reading and writing `cfgs` from different threads, so we should use ConcurrentMap instead of HashMap for this. But in this case, it is not rational (compared to a couple of volatile reads/writes of object reference).